src/net_processing.cpp

```diff
@@ -4246,6 +4246,11 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
                             if (ret.second) {
                                 vRelayExpiration.push_back(std::make_pair(nNow + std::chrono::microseconds{RELAY_TX_CACHE_TIME}.count(), ret.first));
                             }
+                            // Add wtxid-based lookup into mapRelay as well, so that peers can request by wtxid
+                            auto ret2 = mapRelay.emplace(ret.first->second->GetWitnessHash(), ret.first->second);
```

Comment:
confirmed that  not exists tx tx2. txid(tx) == wtxid(tx2) && tx != tx2

```
[7/31/20 14:09] <jeremyrubin> sipa: is this proven: not exists tx tx2. txid(tx) == wtxid(tx2) && tx != tx2
[7/31/20 14:10] <sipa> jeremyrubin: yes, assuming double-sha256 collision resistance that statement is equivalent to ser_without_witness(tx) == ser_with_witness(tx2)
[7/31/20 14:11] <sipa> either both tx and tx2 don't have a witness, and the statement is obviously false
[7/31/20 14:11] <sipa> or tx2 has a witness, in which case the serialization definitely differs, due to the flag byte
[7/31/20 14:13] <jeremyrubin> but it's not possible in tx2 that the flag + marker can be read as the length field for the txins?
[7/31/20 14:13] <jeremyrubin> because a leading 0 byte means it can't be a length right?
[7/31/20 14:13] <jeremyrubin> in which case it would be marker and not flag that guarantees uniqueness?
[7/31/20 14:15] <sipa> a valid transactions cannot have 0 inputs
[7/31/20 14:15] <sipa> the witness serialization is intended to be unambiguously different from the old serialization
[7/31/20 14:15] <sipa> (except that we later discovered that people do pass around transactions with 0 inputs, where things break... but that's not relevant for valid network transactions)(
[7/31/20 14:17] <jeremyrubin> gotcha. and the unambiguity is because it's a compactsize field preceding the vIn
[7/31/20 14:18] <jeremyrubin> so 0 is always read as a 1 byte number in any context
[7/31/20 14:18] <-- proofofkeags (~proofofke@174-29-8-246.hlrn.qwest.net) has left this server (Remote host closed the connection).
[7/31/20 14:19] <sipa> indeed
```